// @stytch/core/dist/testing/index.esm.js
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
import { _ as __awaiter } from '../tslib.es6-f6374340.js';
var MOCK_CAPTCHA = 'mock-captcha-token';
var MOCK_DFP_TELEMTRY_ID = 'dfp-telemetry-id-123';
// jest.setup.js mocks out crypto.getRandomValues to ensure these stay in sync
var MOCK_VERIFIER = '00000000000000000000000000000000';
var MOCK_CHALLENGE = 'hODA6vqpWjTCk_J4rFLkXOU3urXnUqAOaVmhOuEDtlo';
var MOCK_KEYPAIR = {
  code_challenge: MOCK_CHALLENGE,
  code_verifier: MOCK_VERIFIER
};
var MockDFPProtectedAuth = /*#__PURE__*/_createClass(function MockDFPProtectedAuth(enabled, mockTelemetryID) {
  var _this = this;
  var executeRecaptcha = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : function () {
    return Promise.resolve(undefined);
  };
  _classCallCheck(this, MockDFPProtectedAuth);
  this.isEnabled = function () {
    return __awaiter(_this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee() {
      return _regenerator().w(function (_context) {
        while (1) switch (_context.n) {
          case 0:
            return _context.a(2, Promise.resolve(this.state.enabled));
        }
      }, _callee, this);
    }));
  };
  this.getTelemetryID = function () {
    return __awaiter(_this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee2() {
      return _regenerator().w(function (_context2) {
        while (1) switch (_context2.n) {
          case 0:
            return _context2.a(2, Promise.resolve(this.state.mockTelemetryID));
        }
      }, _callee2, this);
    }));
  };
  this.getDFPTelemetryIDAndCaptcha = function () {
    return __awaiter(_this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee3() {
      var _yield$this$state, enabled, executeRecaptcha, dfp_telemetry_id, captcha_token;
      return _regenerator().w(function (_context3) {
        while (1) switch (_context3.n) {
          case 0:
            _context3.n = 1;
            return this.state;
          case 1:
            _yield$this$state = _context3.v;
            enabled = _yield$this$state.enabled;
            executeRecaptcha = _yield$this$state.executeRecaptcha;
            dfp_telemetry_id = undefined;
            captcha_token = undefined;
            if (!enabled) {
              _context3.n = 3;
              break;
            }
            _context3.n = 2;
            return this.getTelemetryID();
          case 2:
            dfp_telemetry_id = _context3.v;
            _context3.n = 5;
            break;
          case 3:
            _context3.n = 4;
            return executeRecaptcha();
          case 4:
            captcha_token = _context3.v;
          case 5:
            return _context3.a(2, {
              dfp_telemetry_id: dfp_telemetry_id,
              captcha_token: captcha_token
            });
        }
      }, _callee3, this);
    }));
  };
  this.executeRecaptcha = function () {
    return __awaiter(_this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee4() {
      return _regenerator().w(function (_context4) {
        while (1) switch (_context4.n) {
          case 0:
            return _context4.a(2, this.executeRecaptcha());
        }
      }, _callee4, this);
    }));
  };
  this.retryWithCaptchaAndDFP = function (e, req) {
    return __awaiter(_this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee5() {
      return _regenerator().w(function (_context5) {
        while (1) switch (_context5.n) {
          case 0:
            if (!this.state.enabled) {
              _context5.n = 3;
              break;
            }
            if (!req.body) {
              _context5.n = 2;
              break;
            }
            req.body.dfp_telemetry_id = this.state.mockTelemetryID;
            _context5.n = 1;
            return this.executeRecaptcha();
          case 1:
            req.body.captcha_token = _context5.v;
          case 2:
            return _context5.a(2, req);
          case 3:
            return _context5.a(2, req);
        }
      }, _callee5, this);
    }));
  };
  this.state = {
    enabled: enabled,
    mockTelemetryID: mockTelemetryID,
    executeRecaptcha: executeRecaptcha
  };
});
var MockDFPProtectedAuthDisabled = function MockDFPProtectedAuthDisabled() {
  return new MockDFPProtectedAuth(false, undefined, function () {
    return Promise.resolve(undefined);
  });
};
var MockDFPProtectedAuthCaptchaOnly = function MockDFPProtectedAuthCaptchaOnly() {
  return new MockDFPProtectedAuth(false, undefined, function () {
    return Promise.resolve(MOCK_CAPTCHA);
  });
};
var MockDFPProtectedAuthDFPAndCaptcha = function MockDFPProtectedAuthDFPAndCaptcha() {
  return new MockDFPProtectedAuth(true, MOCK_DFP_TELEMTRY_ID, function () {
    return Promise.resolve(MOCK_CAPTCHA);
  });
};
var MockDFPProtectedAuthDFPOnly = function MockDFPProtectedAuthDFPOnly() {
  return new MockDFPProtectedAuth(true, MOCK_DFP_TELEMTRY_ID, function () {
    return Promise.resolve(undefined);
  });
};
var MockPKCEManager = /*#__PURE__*/function () {
  function MockPKCEManager() {
    _classCallCheck(this, MockPKCEManager);
    this.pair = undefined;
  }
  return _createClass(MockPKCEManager, [{
    key: "setPKPair",
    value: function setPKPair(pair) {
      this.pair = pair;
    }
  }, {
    key: "clearPKPair",
    value: function clearPKPair() {
      this.pair = undefined;
    }
  }, {
    key: "getPKPair",
    value: function getPKPair() {
      return this.pair;
    }
  }, {
    key: "startPKCETransaction",
    value: function startPKCETransaction() {
      this.pair = MOCK_KEYPAIR;
      return Promise.resolve(this.pair);
    }
  }]);
}();
var createTestFixtures = function createTestFixtures() {
  var networkClient = {
    fetchSDK: jest.fn(),
    retriableFetchSDK: jest.fn(),
    logEvent: jest.fn(),
    createTelemetryBlob: jest.fn(),
    updateSessionToken: jest.fn()
  };
  var apiNetworkClient = {
    fetchSDK: jest.fn(),
    retriableFetchSDK: jest.fn(),
    logEvent: jest.fn(),
    createTelemetryBlob: jest.fn(),
    updateSessionToken: jest.fn()
  };
  var updateSession = jest.fn();
  var subscriptionService = {
    updateStateAndTokens: jest.fn(),
    updateState: jest.fn(),
    getState: jest.fn(),
    updateUser: jest.fn(),
    destroyState: jest.fn(),
    updateSession: updateSession,
    withUpdateSession: jest.fn().mockImplementation(function (fn) {
      return function () {
        for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
          args[_key] = arguments[_key];
        }
        return __awaiter(void 0, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee6() {
          var resp;
          return _regenerator().w(function (_context6) {
            while (1) switch (_context6.n) {
              case 0:
                _context6.n = 1;
                return fn.apply(void 0, args);
              case 1:
                resp = _context6.v;
                updateSession(resp);
                return _context6.a(2, resp);
            }
          }, _callee6);
        }));
      };
    }),
    destroySession: jest.fn(),
    getSession: jest.fn(),
    subscribeToState: jest.fn(),
    getUser: jest.fn(),
    getTokens: jest.fn(),
    syncFromDeviceStorage: jest.fn(),
    updateTokens: jest.fn(),
    getIntermediateSessionToken: jest.fn(),
    getFromCache: jest.fn(),
    setCacheRefreshed: jest.fn()
  };
  var pkceManager = new MockPKCEManager();
  var secondPKCEManager = new MockPKCEManager();
  var captchaProvider = function captchaProvider() {
    return Promise.resolve(MOCK_CAPTCHA);
  };
  var dfpProtectedAuth = new MockDFPProtectedAuth(false, undefined, captchaProvider);
  return {
    networkClient: networkClient,
    apiNetworkClient: apiNetworkClient,
    subscriptionService: subscriptionService,
    pkceManager: pkceManager,
    secondPKCEManager: secondPKCEManager,
    captchaProvider: captchaProvider,
    dfpProtectedAuth: dfpProtectedAuth
  };
};
var createB2BTestFixtures = function createB2BTestFixtures() {
  var networkClient = {
    fetchSDK: jest.fn(),
    retriableFetchSDK: jest.fn(),
    logEvent: jest.fn(),
    createTelemetryBlob: jest.fn(),
    updateSessionToken: jest.fn()
  };
  var apiNetworkClient = {
    fetchSDK: jest.fn(),
    retriableFetchSDK: jest.fn(),
    logEvent: jest.fn(),
    createTelemetryBlob: jest.fn(),
    updateSessionToken: jest.fn()
  };
  var updateSession = jest.fn();
  var subscriptionService = {
    updateStateAndTokens: jest.fn(),
    updateState: jest.fn(),
    getState: jest.fn(),
    updateMember: jest.fn(),
    destroyState: jest.fn(),
    updateSession: updateSession,
    withUpdateSession: jest.fn().mockImplementation(function (fn) {
      return function () {
        for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
          args[_key2] = arguments[_key2];
        }
        return __awaiter(void 0, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee7() {
          var resp;
          return _regenerator().w(function (_context7) {
            while (1) switch (_context7.n) {
              case 0:
                _context7.n = 1;
                return fn.apply(void 0, args);
              case 1:
                resp = _context7.v;
                updateSession(resp);
                return _context7.a(2, resp);
            }
          }, _callee7);
        }));
      };
    }),
    destroySession: jest.fn(),
    getSession: jest.fn(),
    subscribeToState: jest.fn(),
    getMember: jest.fn(),
    getTokens: jest.fn(),
    syncFromDeviceStorage: jest.fn(),
    updateTokens: jest.fn(),
    getIntermediateSessionToken: jest.fn(),
    getOrganization: jest.fn(),
    updateOrganization: jest.fn(),
    getFromCache: jest.fn(),
    setCacheRefreshed: jest.fn()
  };
  var pkceManager = new MockPKCEManager();
  var secondPKCEManager = new MockPKCEManager();
  var captchaProvider = function captchaProvider() {
    return Promise.resolve(MOCK_CAPTCHA);
  };
  return {
    networkClient: networkClient,
    apiNetworkClient: apiNetworkClient,
    subscriptionService: subscriptionService,
    pkceManager: pkceManager,
    secondPKCEManager: secondPKCEManager,
    captchaProvider: captchaProvider
  };
};
var createResolvablePromise = function createResolvablePromise() {
  var resolve;
  var reject;
  var promise = new Promise(function (res, rej) {
    resolve = res;
    reject = rej;
  });
  return {
    promise: promise,
    resolve: resolve,
    reject: reject
  };
};
function getMockStorageClient() {
  return {
    getData: jest.fn(),
    setData: jest.fn(),
    clearData: jest.fn()
  };
}
var PUBLIC_TOKEN = 'public-token-test-1234';
export { MockDFPProtectedAuth, MockDFPProtectedAuthCaptchaOnly, MockDFPProtectedAuthDFPAndCaptcha, MockDFPProtectedAuthDFPOnly, MockDFPProtectedAuthDisabled, MockPKCEManager, PUBLIC_TOKEN, createB2BTestFixtures, createResolvablePromise, createTestFixtures, getMockStorageClient };
