// @stytch/vanilla-js/dist/index.headless-Dj4KRdHg.js
'use strict';

function _callSuper(t, o, e) { return o = _getPrototypeOf(o), _possibleConstructorReturn(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], _getPrototypeOf(t).constructor) : o.apply(t, e)); }
function _possibleConstructorReturn(t, e) { if (e && ("object" == _typeof(e) || "function" == typeof e)) return e; if (void 0 !== e) throw new TypeError("Derived constructors may only return object or undefined"); return _assertThisInitialized(t); }
function _assertThisInitialized(e) { if (void 0 === e) throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); return e; }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function _getPrototypeOf(t) { return _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function (t) { return t.__proto__ || Object.getPrototypeOf(t); }, _getPrototypeOf(t); }
function _inherits(t, e) { if ("function" != typeof e && null !== e) throw new TypeError("Super expression must either be null or a function"); t.prototype = Object.create(e && e.prototype, { constructor: { value: t, writable: !0, configurable: !0 } }), Object.defineProperty(t, "prototype", { writable: !1 }), e && _setPrototypeOf(t, e); }
function _setPrototypeOf(t, e) { return _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function (t, e) { return t.__proto__ = e, t; }, _setPrototypeOf(t, e); }
function _slicedToArray(r, e) { return _arrayWithHoles(r) || _iterableToArrayLimit(r, e) || _unsupportedIterableToArray(r, e) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(r) { if (Array.isArray(r)) return r; }
function _toConsumableArray(r) { return _arrayWithoutHoles(r) || _iterableToArray(r) || _unsupportedIterableToArray(r) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _iterableToArray(r) { if ("undefined" != typeof Symbol && null != r[Symbol.iterator] || null != r["@@iterator"]) return Array.from(r); }
function _arrayWithoutHoles(r) { if (Array.isArray(r)) return _arrayLikeToArray(r); }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; }
function _regenerator() { /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */ var e, t, r = "function" == typeof Symbol ? Symbol : {}, n = r.iterator || "@@iterator", o = r.toStringTag || "@@toStringTag"; function i(r, n, o, i) { var c = n && n.prototype instanceof Generator ? n : Generator, u = Object.create(c.prototype); return _regeneratorDefine2(u, "_invoke", function (r, n, o) { var i, c, u, f = 0, p = o || [], y = !1, G = { p: 0, n: 0, v: e, a: d, f: d.bind(e, 4), d: function d(t, r) { return i = t, c = 0, u = e, G.n = r, a; } }; function d(r, n) { for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) { var o, i = p[t], d = G.p, l = i[2]; r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0)); } if (o || r > 1) return a; throw y = !0, n; } return function (o, p, l) { if (f > 1) throw TypeError("Generator is already running"); for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) { i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u); try { if (f = 2, i) { if (c || (o = "next"), t = i[o]) { if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object"); if (!t.done) return t; u = t.value, c < 2 && (c = 0); } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1); i = e; } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break; } catch (t) { i = e, c = 1, u = t; } finally { f = 1; } } return { value: t, done: y }; }; }(r, o, i), !0), u; } var a = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} t = Object.getPrototypeOf; var c = [][n] ? t(t([][n]())) : (_regeneratorDefine2(t = {}, n, function () { return this; }), t), u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c); function f(e) { return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, _regeneratorDefine2(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, _regeneratorDefine2(u, "constructor", GeneratorFunctionPrototype), _regeneratorDefine2(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", _regeneratorDefine2(GeneratorFunctionPrototype, o, "GeneratorFunction"), _regeneratorDefine2(u), _regeneratorDefine2(u, o, "Generator"), _regeneratorDefine2(u, n, function () { return this; }), _regeneratorDefine2(u, "toString", function () { return "[object Generator]"; }), (_regenerator = function _regenerator() { return { w: i, m: f }; })(); }
function _regeneratorDefine2(e, r, n, t) { var i = Object.defineProperty; try { i({}, "", {}); } catch (e) { i = 0; } _regeneratorDefine2 = function _regeneratorDefine(e, r, n, t) { function o(r, n) { _regeneratorDefine2(e, r, function (e) { return this._invoke(r, n, e); }); } r ? i ? i(e, r, { value: n, enumerable: !t, configurable: !t, writable: !t }) : e[r] = n : (o("next", 0), o("throw", 1), o("return", 2)); }, _regeneratorDefine2(e, r, n, t); }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
var internal = require('./internal-BWEZsr7G.js');
var BOOTSTRAP_CACHE_KEY = 'bootstrap';
var DEFAULT_BOOTSTRAP = function DEFAULT_BOOTSTRAP() {
  return {
    displayWatermark: false,
    projectName: null,
    cnameDomain: null,
    emailDomains: ['stytch.com'],
    captchaSettings: {
      enabled: false
    },
    pkceRequiredForEmailMagicLinks: false,
    pkceRequiredForPasswordResets: false,
    pkceRequiredForOAuth: false,
    pkceRequiredForSso: false,
    slugPattern: null,
    createOrganizationEnabled: false,
    passwordConfig: null,
    runDFPProtectedAuth: false,
    rbacPolicy: null,
    siweRequiredForCryptoWallets: false,
    vertical: null
  };
};
var BootstrapDataManager = /*#__PURE__*/function () {
  function BootstrapDataManager(_publicToken, _networkClient, _dataLayer) {
    var _this = this;
    _classCallCheck(this, BootstrapDataManager);
    this._publicToken = _publicToken;
    this._networkClient = _networkClient;
    this._dataLayer = _dataLayer;
    this._bootstrapDataPromise = this._networkClient.fetchSDK({
      url: "/projects/bootstrap/".concat(this._publicToken),
      method: 'GET'
    }).then(BootstrapDataManager.mapBootstrapData).then(function (data) {
      _this._dataLayer.setItem(BOOTSTRAP_CACHE_KEY, JSON.stringify(data));
      return data;
    })["catch"](function (error) {
      internal.logger.error(error);
      return DEFAULT_BOOTSTRAP();
    });
  }
  return _createClass(BootstrapDataManager, [{
    key: "getSync",
    value: function getSync() {
      var cached = this._dataLayer.getItem(BOOTSTRAP_CACHE_KEY);
      if (cached === null) {
        return DEFAULT_BOOTSTRAP();
      }
      try {
        return JSON.parse(cached);
      } catch (_a) {
        return DEFAULT_BOOTSTRAP();
      }
    }
  }, {
    key: "getAsync",
    value: function getAsync() {
      return this._bootstrapDataPromise;
    }
  }], [{
    key: "mapBootstrapData",
    value: function mapBootstrapData(response) {
      var _a, _b;
      var passwordConfig = response.password_config !== null ? {
        ludsComplexity: response.password_config.luds_complexity,
        ludsMinimumCount: response.password_config.luds_minimum_count
      } : null;
      return {
        projectName: response.project_name,
        displayWatermark: !response.disable_sdk_watermark,
        captchaSettings: response.captcha_settings,
        cnameDomain: response.cname_domain,
        emailDomains: response.email_domains,
        pkceRequiredForEmailMagicLinks: response.pkce_required_for_email_magic_links,
        pkceRequiredForPasswordResets: response.pkce_required_for_password_resets,
        pkceRequiredForOAuth: response.pkce_required_for_oauth,
        pkceRequiredForSso: response.pkce_required_for_sso,
        slugPattern: response.slug_pattern,
        createOrganizationEnabled: response.create_organization_enabled,
        passwordConfig: passwordConfig,
        runDFPProtectedAuth: (_a = response.dfp_protected_auth_enabled) !== null && _a !== void 0 ? _a : false,
        dfpProtectedAuthMode: response.dfp_protected_auth_mode,
        rbacPolicy: (_b = response.rbac_policy) !== null && _b !== void 0 ? _b : null,
        siweRequiredForCryptoWallets: response.siwe_required_for_crypto_wallets,
        vertical: response.vertical
      };
    }
  }]);
}();
var loadRecaptchaClient = function loadRecaptchaClient(siteKey) {
  return internal.loadESModule("https://www.google.com/recaptcha/enterprise.js?render=".concat(siteKey), function () {
    return window.grecaptcha.enterprise;
  });
};
var CaptchaProvider = /*#__PURE__*/_createClass(function CaptchaProvider(bootstrapPromise) {
  var _this2 = this;
  _classCallCheck(this, CaptchaProvider);
  this.bootstrapPromise = bootstrapPromise;
  this.executeRecaptcha = function () {
    return internal.__awaiter(_this2, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee() {
      var _yield$this$state, captchaClient, configured, siteKey;
      return _regenerator().w(function (_context) {
        while (1) switch (_context.n) {
          case 0:
            _context.n = 1;
            return this.state;
          case 1:
            _yield$this$state = _context.v;
            captchaClient = _yield$this$state.captchaClient;
            configured = _yield$this$state.configured;
            siteKey = _yield$this$state.siteKey;
            if (configured) {
              _context.n = 2;
              break;
            }
            return _context.a(2, undefined);
          case 2:
            _context.n = 3;
            return new Promise(function (resolve) {
              return captchaClient.ready(resolve);
            });
          case 3:
            return _context.a(2, captchaClient.execute(siteKey, {
              action: 'LOGIN'
            }));
        }
      }, _callee, this);
    }));
  };
  this.state = bootstrapPromise.then(function (bootstrapData) {
    return internal.__awaiter(_this2, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee2() {
      var _t, _t2;
      return _regenerator().w(function (_context2) {
        while (1) switch (_context2.n) {
          case 0:
            if (bootstrapData.captchaSettings.enabled) {
              _context2.n = 1;
              break;
            }
            return _context2.a(2, {
              configured: false
            });
          case 1:
            _context2.n = 2;
            return loadRecaptchaClient(bootstrapData.captchaSettings.siteKey);
          case 2:
            _t = _context2.v;
            _t2 = bootstrapData.captchaSettings.siteKey;
            return _context2.a(2, {
              configured: true,
              captchaClient: _t,
              siteKey: _t2
            });
        }
      }, _callee2);
    }));
  });
});
var ClientsideServicesProvider = /*#__PURE__*/function () {
  function ClientsideServicesProvider(iframeURL) {
    var _this3 = this;
    _classCallCheck(this, ClientsideServicesProvider);
    this.iframeURL = iframeURL;
    this.oneTapStart = function (req) {
      return _this3.call('oneTapStart', req);
    };
    this.oneTapSubmit = function (req) {
      return _this3.call('oneTapSubmit', req);
    };
    this.parsedPhoneNumber = function (req) {
      return _this3.call('parsedPhoneNumber', req);
    };
  }
  return _createClass(ClientsideServicesProvider, [{
    key: "frameClient",
    get: function get() {
      var _a;
      this._frameClient = (_a = this._frameClient) !== null && _a !== void 0 ? _a : new internal.IframeHostClient(this.iframeURL);
      return this._frameClient;
    }
  }, {
    key: "call",
    value: function call(handlerName, req) {
      return this.frameClient.call(handlerName, [req]);
    }
  }]);
}();
var version = "5.44.2";
var NetworkClient = /*#__PURE__*/function () {
  function NetworkClient(_publicToken, _subscriptionDataLayer, _liveAPIURL, _testAPIURL, additionalTelemetryDataFn) {
    _classCallCheck(this, NetworkClient);
    this._publicToken = _publicToken;
    this._subscriptionDataLayer = _subscriptionDataLayer;
    this.additionalTelemetryDataFn = additionalTelemetryDataFn;
    this.updateSessionToken = function () {
      return null;
    };
    this.baseURL = _liveAPIURL;
    if (internal.isTestPublicToken(_publicToken)) {
      this.baseURL = _testAPIURL;
    }
    this.eventLogger = new internal.EventLogger({
      maxBatchSize: internal.DEFAULT_MAX_BATCH_SIZE,
      intervalDurationMs: internal.DEFAULT_INTERVAL_DURATION_MS,
      logEventURL: this.buildSDKUrl('/events')
    });
  }
  return _createClass(NetworkClient, [{
    key: "logEvent",
    value: function logEvent(_ref) {
      var name = _ref.name,
        details = _ref.details,
        _ref$error = _ref.error,
        error = _ref$error === void 0 ? {} : _ref$error;
      this.eventLogger.logEvent(this.createTelemetryBlob(), {
        public_token: this._publicToken,
        event_name: name,
        details: details,
        error_code: error.error_code,
        error_description: error.error_description,
        http_status_code: error.http_status_code
      });
    }
  }, {
    key: "createTelemetryBlob",
    value: function createTelemetryBlob() {
      return Object.assign(Object.assign({
        event_id: internal.createEventId(),
        app_session_id: internal.createAppSessionId(),
        persistent_id: internal.createPersistentId(),
        client_sent_at: new Date().toISOString(),
        timezone: Intl.DateTimeFormat().resolvedOptions().timeZone
      }, this.additionalTelemetryDataFn()), {
        app: {
          identifier: window.location.hostname
        },
        sdk: {
          identifier: 'Stytch.js Javascript SDK',
          version: version
        }
      });
    }
  }, {
    key: "fetchSDK",
    value: function fetchSDK(_ref2) {
      var url = _ref2.url,
        body = _ref2.body,
        method = _ref2.method;
      return internal.__awaiter(this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee3() {
        var sessionToken, basicAuthHeader, xSDKClientHeader, xSDKParentHostHeader;
        return _regenerator().w(function (_context3) {
          while (1) switch (_context3.n) {
            case 0:
              sessionToken = this._subscriptionDataLayer.readSessionCookie().session_token;
              basicAuthHeader = 'Basic ' + window.btoa(this._publicToken + ':' + (sessionToken || this._publicToken));
              xSDKClientHeader = window.btoa(JSON.stringify(this.createTelemetryBlob()));
              xSDKParentHostHeader = window.location.origin;
              return _context3.a(2, internal.baseFetchSDK({
                basicAuthHeader: basicAuthHeader,
                body: body,
                finalURL: this.buildSDKUrl(url),
                method: method,
                xSDKClientHeader: xSDKClientHeader,
                xSDKParentHostHeader: xSDKParentHostHeader
              }));
          }
        }, _callee3, this);
      }));
    }
  }, {
    key: "submitFormSDK",
    value: function submitFormSDK(_ref3) {
      var url = _ref3.url,
        body = _ref3.body,
        method = _ref3.method;
      return internal.__awaiter(this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee4() {
        var sessionToken, basicAuthHeader, xSDKClientHeader, xSDKParentHostHeader;
        return _regenerator().w(function (_context4) {
          while (1) switch (_context4.n) {
            case 0:
              sessionToken = this._subscriptionDataLayer.readSessionCookie().session_token;
              basicAuthHeader = 'Basic ' + window.btoa(this._publicToken + ':' + (sessionToken || this._publicToken));
              xSDKClientHeader = window.btoa(JSON.stringify(this.createTelemetryBlob()));
              xSDKParentHostHeader = window.location.origin;
              return _context4.a(2, internal.baseSubmitFormSDK({
                basicAuthHeader: basicAuthHeader,
                body: body,
                finalURL: this.buildSDKUrl(url),
                method: method,
                xSDKClientHeader: xSDKClientHeader,
                xSDKParentHostHeader: xSDKParentHostHeader
              }));
          }
        }, _callee4, this);
      }));
    }
  }, {
    key: "retriableFetchSDK",
    value: function retriableFetchSDK(_ref4) {
      var url = _ref4.url,
        body = _ref4.body,
        method = _ref4.method,
        retryCallback = _ref4.retryCallback;
      return internal.__awaiter(this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee5() {
        var sessionToken, basicAuthHeader, xSDKClientHeader, xSDKParentHostHeader;
        return _regenerator().w(function (_context5) {
          while (1) switch (_context5.n) {
            case 0:
              sessionToken = this._subscriptionDataLayer.readSessionCookie().session_token;
              basicAuthHeader = 'Basic ' + window.btoa(this._publicToken + ':' + (sessionToken || this._publicToken));
              xSDKClientHeader = window.btoa(JSON.stringify(this.createTelemetryBlob()));
              xSDKParentHostHeader = window.location.origin;
              return _context5.a(2, internal.retriableFetchSDK({
                basicAuthHeader: basicAuthHeader,
                body: body,
                finalURL: this.buildSDKUrl(url),
                method: method,
                xSDKClientHeader: xSDKClientHeader,
                xSDKParentHostHeader: xSDKParentHostHeader,
                retryCallback: retryCallback
              }));
          }
        }, _callee5, this);
      }));
    }
  }, {
    key: "buildSDKUrl",
    value: function buildSDKUrl(url) {
      return "".concat(this.baseURL, "/sdk/v1").concat(url);
    }
  }]);
}();
var PKCE_VERIFIER_STORAGE_KEY = 'PKCE_VERIFIER';
function toHex(n) {
  var str = n.toString(16);
  if (str.length === 1) {
    str = '0' + str;
  }
  return str;
}
function base64URLEncode(buf) {
  var _String$fromCharCode;
  return btoa((_String$fromCharCode = String.fromCharCode).call.apply(_String$fromCharCode, [null].concat(_toConsumableArray(new Uint8Array(buf))))).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}
var PKCEManager = /*#__PURE__*/function () {
  function PKCEManager(_dataLayer, namespace) {
    _classCallCheck(this, PKCEManager);
    this._dataLayer = _dataLayer;
    this.namespace = namespace;
  }
  return _createClass(PKCEManager, [{
    key: "key",
    value: function key() {
      return "".concat(PKCE_VERIFIER_STORAGE_KEY, ":").concat(this.namespace);
    }
  }, {
    key: "startPKCETransaction",
    value: function startPKCETransaction() {
      return internal.__awaiter(this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee6() {
        var keyPair;
        return _regenerator().w(function (_context6) {
          while (1) switch (_context6.n) {
            case 0:
              _context6.n = 1;
              return PKCEManager.createProofkeyPair();
            case 1:
              keyPair = _context6.v;
              this._dataLayer.setItem(this.key(), JSON.stringify(keyPair));
              return _context6.a(2, keyPair);
          }
        }, _callee6, this);
      }));
    }
  }, {
    key: "getPKPair",
    value: function getPKPair() {
      var serialized = this._dataLayer.getItem(this.key());
      if (serialized === null) {
        return undefined;
      }
      try {
        return JSON.parse(serialized);
      } catch (_a) {
        internal.logger.warn('Found malformed Proof Key pair in localstorage.');
        return undefined;
      }
    }
  }, {
    key: "clearPKPair",
    value: function clearPKPair() {
      return this._dataLayer.removeItem(this.key());
    }
  }], [{
    key: "createProofkeyPair",
    value: function createProofkeyPair() {
      return internal.__awaiter(this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee7() {
        var bytes, codeVerifier, digest;
        return _regenerator().w(function (_context7) {
          while (1) switch (_context7.n) {
            case 0:
              bytes = new Uint32Array(16);
              window.crypto.getRandomValues(bytes);
              codeVerifier = Array.from(bytes).map(toHex).join('');
              _context7.n = 1;
              return window.crypto.subtle.digest('SHA-256', new TextEncoder().encode(codeVerifier));
            case 1:
              digest = _context7.v;
              return _context7.a(2, {
                code_challenge: base64URLEncode(digest),
                code_verifier: codeVerifier
              });
          }
        }, _callee7);
      }));
    }
  }]);
}();
/*! js-cookie v3.0.1 | MIT */
function assign(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i];
    for (var key in source) {
      target[key] = source[key];
    }
  }
  return target;
}
var defaultConverter = {
  read: function read(value) {
    if (value[0] === '"') {
      value = value.slice(1, -1);
    }
    return value.replace(/(%[\dA-F]{2})+/gi, decodeURIComponent);
  },
  write: function write(value) {
    return encodeURIComponent(value).replace(/%(2[346BF]|3[AC-F]|40|5[BDE]|60|7[BCD])/g, decodeURIComponent);
  }
};
function init(converter, defaultAttributes) {
  function set(key, value, attributes) {
    if (typeof document === 'undefined') {
      return;
    }
    attributes = assign({}, defaultAttributes, attributes);
    if (typeof attributes.expires === 'number') {
      attributes.expires = new Date(Date.now() + attributes.expires * 864e5);
    }
    if (attributes.expires) {
      attributes.expires = attributes.expires.toUTCString();
    }
    key = encodeURIComponent(key).replace(/%(2[346B]|5E|60|7C)/g, decodeURIComponent).replace(/[()]/g, escape);
    var stringifiedAttributes = '';
    for (var attributeName in attributes) {
      if (!attributes[attributeName]) {
        continue;
      }
      stringifiedAttributes += '; ' + attributeName;
      if (attributes[attributeName] === true) {
        continue;
      }
      stringifiedAttributes += '=' + attributes[attributeName].split(';')[0];
    }
    return document.cookie = key + '=' + converter.write(value, key) + stringifiedAttributes;
  }
  function get(key) {
    if (typeof document === 'undefined' || arguments.length && !key) {
      return;
    }
    var cookies = document.cookie ? document.cookie.split('; ') : [];
    var jar = {};
    for (var i = 0; i < cookies.length; i++) {
      var parts = cookies[i].split('=');
      var value = parts.slice(1).join('=');
      try {
        var foundKey = decodeURIComponent(parts[0]);
        jar[foundKey] = converter.read(value, foundKey);
        if (key === foundKey) {
          break;
        }
      } catch (e) {}
    }
    return key ? jar[key] : jar;
  }
  return Object.create({
    set: set,
    get: get,
    remove: function remove(key, attributes) {
      set(key, '', assign({}, attributes, {
        expires: -1
      }));
    },
    withAttributes: function withAttributes(attributes) {
      return init(this.converter, assign({}, this.attributes, attributes));
    },
    withConverter: function withConverter(converter) {
      return init(assign({}, this.converter, converter), this.attributes);
    }
  }, {
    attributes: {
      value: Object.freeze(defaultAttributes)
    },
    converter: {
      value: Object.freeze(converter)
    }
  });
}
var api = init(defaultConverter, {
  path: '/'
});
var createDeepEqual = function createDeepEqual() {
  var _ref5 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
    _ref5$KEYS_TO_EXCLUDE = _ref5.KEYS_TO_EXCLUDE,
    KEYS_TO_EXCLUDE = _ref5$KEYS_TO_EXCLUDE === void 0 ? [] : _ref5$KEYS_TO_EXCLUDE;
  var _deepEqual = function deepEqual(a, b) {
    if (_typeof(a) !== _typeof(b)) return false;
    if (a === null || b === null) return a === b;
    if (_typeof(a) === 'object') {
      if (Object.keys(a).length !== Object.keys(b).length || Object.keys(a).some(function (k) {
        return !(k in b);
      })) return false;
      return Object.entries(a).filter(function (_ref6) {
        var _ref7 = _slicedToArray(_ref6, 1),
          k = _ref7[0];
        return !KEYS_TO_EXCLUDE.includes(k);
      }).every(function (_ref8) {
        var _ref9 = _slicedToArray(_ref8, 2),
          k = _ref9[0],
          v = _ref9[1];
        return _deepEqual(v, b[k]);
      });
    }
    return a === b;
  };
  return _deepEqual;
};
var STYTCH_SESSION_COOKIE = 'stytch_session';
var STYTCH_SESSION_JWT_COOKIE = 'stytch_session_jwt';
var STYTCH_INTERMEDIATE_SESSION_TOKEN_COOKIE = 'stytch_intermediate_session_token';
var SEEN_DOMAINS_KEY = 'seen_domains';
var SubscriptionDataLayer = /*#__PURE__*/function () {
  function SubscriptionDataLayer(publicToken, options) {
    _classCallCheck(this, SubscriptionDataLayer);
    this._opaqueTokenCookieName = null;
    this._jwtCookieName = null;
    this._cookiePath = null;
    this._domain = null;
    this._cookieAvailableToSubdomains = false;
    this._istCookieName = null;
    this.publicToken = publicToken;
    this.state = null;
    this.stateKeysUpdated = new Set();
    this.subscriptions = {};
    this._localStorage = internal.getKeyBoundStorage(internal.safeLocalStorage, publicToken);
    this.browserSessionStorage = internal.getKeyBoundStorage(internal.safeSessionStorage, publicToken);
    if (options === null || options === void 0 ? void 0 : options.cookieOptions) {
      this._opaqueTokenCookieName = options.cookieOptions.opaqueTokenCookieName || null;
      this._jwtCookieName = options.cookieOptions.jwtCookieName || null;
      this._cookiePath = options.cookieOptions.path || null;
      this._domain = options.cookieOptions.domain || null;
      this._cookieAvailableToSubdomains = options.cookieOptions.availableToSubdomains || false;
      this._istCookieName = options.cookieOptions.istCookieName || null;
    }
    var localStorageState = this._localStorage.getItem('');
    if (!localStorageState) {
      return;
    }
    var parsedState;
    try {
      parsedState = JSON.parse(localStorageState);
    } catch (_a) {
      this.syncToLocalStorage();
      return;
    }
    this.state = parsedState;
  }
  return _createClass(SubscriptionDataLayer, [{
    key: "opaqueTokenCookieName",
    get: function get() {
      var _a;
      return (_a = this._opaqueTokenCookieName) !== null && _a !== void 0 ? _a : STYTCH_SESSION_COOKIE;
    }
  }, {
    key: "jwtCookieName",
    get: function get() {
      var _a;
      return (_a = this._jwtCookieName) !== null && _a !== void 0 ? _a : STYTCH_SESSION_JWT_COOKIE;
    }
  }, {
    key: "istCookieName",
    get: function get() {
      var _a;
      return (_a = this._istCookieName) !== null && _a !== void 0 ? _a : STYTCH_INTERMEDIATE_SESSION_TOKEN_COOKIE;
    }
  }, {
    key: "readSessionCookie",
    value: function readSessionCookie() {
      return {
        session_token: api.get(this.opaqueTokenCookieName),
        session_jwt: api.get(this.jwtCookieName)
      };
    }
  }, {
    key: "readIntermediateSessionTokenCookie",
    value: function readIntermediateSessionTokenCookie() {
      return api.get(this.istCookieName);
    }
  }, {
    key: "writeSessionCookie",
    value: function writeSessionCookie(stateDiff) {
      var _a, _b, _c, _d;
      var state = stateDiff.state,
        session_token = stateDiff.session_token,
        session_jwt = stateDiff.session_jwt;
      var cookieOpts = SubscriptionDataLayer.generateCookieOpts({
        expiresAt: (_b = (_a = state === null || state === void 0 ? void 0 : state.session) === null || _a === void 0 ? void 0 : _a.expires_at) !== null && _b !== void 0 ? _b : '',
        availableToSubdomains: this._cookieAvailableToSubdomains,
        path: this._cookiePath,
        domain: this._domain
      });
      if (cookieOpts.domain) {
        this.addSeenDomain(cookieOpts.domain);
      }
      api.set(this.opaqueTokenCookieName, session_token, cookieOpts);
      api.set(this.jwtCookieName, session_jwt, cookieOpts);
      var alternateCookieOptions = SubscriptionDataLayer.generateCookieOpts({
        expiresAt: (_d = (_c = state === null || state === void 0 ? void 0 : state.session) === null || _c === void 0 ? void 0 : _c.expires_at) !== null && _d !== void 0 ? _d : '',
        availableToSubdomains: !this._cookieAvailableToSubdomains,
        path: this._cookiePath,
        domain: this._domain
      });
      if (alternateCookieOptions.domain) {
        this.addSeenDomain(alternateCookieOptions.domain);
      }
      if (internal.hasMultipleCookies(this.jwtCookieName)) {
        api.remove(this.jwtCookieName, alternateCookieOptions);
      }
      if (internal.hasMultipleCookies(this.opaqueTokenCookieName)) {
        api.remove(this.opaqueTokenCookieName, alternateCookieOptions);
      }
      if (internal.hasMultipleCookies(this.jwtCookieName)) {
        internal.logger.warn('Could not remove extraneous JWT cookie. This might happen if the cookie has been set using multiple `path` settings, and may produce unwanted behavior.');
      }
      if (internal.hasMultipleCookies(this.opaqueTokenCookieName)) {
        internal.logger.warn('Could not remove extraneous opaque token cookie.');
      }
    }
  }, {
    key: "writeIntermediateSessionTokenCookie",
    value: function writeIntermediateSessionTokenCookie(IST) {
      var expiresAtTime = new Date(Date.now() + 10 * 60000);
      var cookieOpts = SubscriptionDataLayer.generateCookieOpts({
        expiresAt: expiresAtTime.toString(),
        availableToSubdomains: this._cookieAvailableToSubdomains,
        path: this._cookiePath,
        domain: this._domain
      });
      api.set(this.istCookieName, IST, cookieOpts);
    }
  }, {
    key: "removeSessionCookie",
    value: function removeSessionCookie() {
      this.removeCookies([this.opaqueTokenCookieName, this.jwtCookieName]);
    }
  }, {
    key: "removeISTCookie",
    value: function removeISTCookie() {
      this.removeCookies([this.istCookieName]);
    }
  }, {
    key: "removeCookies",
    value: function removeCookies(cookiesToRemove) {
      var _this4 = this;
      var trackedDomains = this.getSeenDomains();
      var allDomains = [this._domain, null].concat(_toConsumableArray(trackedDomains));
      var uniqueDomains = _toConsumableArray(new Set(allDomains));
      [true, false].forEach(function (availableToSubdomains) {
        [_this4._cookiePath, null].forEach(function (path) {
          uniqueDomains.forEach(function (domain) {
            var cookieOpts = SubscriptionDataLayer.generateCookieOpts({
              expiresAt: new Date(0).toString(),
              availableToSubdomains: availableToSubdomains,
              path: path,
              domain: domain
            });
            cookiesToRemove.forEach(function (cookieName) {
              api.remove(cookieName, cookieOpts);
            });
          });
        });
      });
    }
  }, {
    key: "syncToLocalStorage",
    value: function syncToLocalStorage() {
      this._localStorage.setItem('', JSON.stringify(this.state));
    }
  }, {
    key: "getItem",
    value: function getItem(key) {
      return this._localStorage.getItem(key);
    }
  }, {
    key: "setItem",
    value: function setItem(key, value) {
      this._localStorage.setItem(key, value);
    }
  }, {
    key: "removeItem",
    value: function removeItem(key) {
      this._localStorage.removeItem(key);
    }
  }, {
    key: "getSeenDomains",
    value: function getSeenDomains() {
      var storedDomains = this.getItem(SEEN_DOMAINS_KEY);
      if (!storedDomains) {
        return [];
      }
      try {
        var parsed = JSON.parse(storedDomains);
        return Array.isArray(parsed) ? parsed : [];
      } catch (_a) {
        return [];
      }
    }
  }, {
    key: "addSeenDomain",
    value: function addSeenDomain(domain) {
      if (!domain) {
        return;
      }
      var storedDomains = this.getSeenDomains();
      if (!storedDomains.includes(domain)) {
        storedDomains.push(domain);
        this.setItem(SEEN_DOMAINS_KEY, JSON.stringify(storedDomains));
      }
    }
  }], [{
    key: "generateCookieOpts",
    value: function generateCookieOpts(_ref0) {
      var path = _ref0.path,
        domain = _ref0.domain,
        availableToSubdomains = _ref0.availableToSubdomains,
        expiresAt = _ref0.expiresAt;
      var cookieOpts = {
        expires: new Date(expiresAt),
        sameSite: 'lax'
      };
      if (path) {
        cookieOpts.path = path;
      }
      if (internal.isLocalhost()) {
        cookieOpts.secure = false;
      } else {
        if (availableToSubdomains) {
          cookieOpts.domain = domain || window.location.host;
        }
        cookieOpts.secure = true;
      }
      return cookieOpts;
    }
  }]);
}();
var B2BSubscriptionDataLayer = /*#__PURE__*/function (_SubscriptionDataLaye) {
  function B2BSubscriptionDataLayer() {
    _classCallCheck(this, B2BSubscriptionDataLayer);
    return _callSuper(this, B2BSubscriptionDataLayer, arguments);
  }
  _inherits(B2BSubscriptionDataLayer, _SubscriptionDataLaye);
  return _createClass(B2BSubscriptionDataLayer);
}(SubscriptionDataLayer);
var b2bDataLayerCacheSymbol = Symbol["for"]('__stytch_b2b_DataLayer');
var getB2BDataLayerCache = function getB2BDataLayerCache() {
  var wdw = window;
  if (!wdw[b2bDataLayerCacheSymbol]) {
    wdw[b2bDataLayerCacheSymbol] = {};
  }
  return wdw[b2bDataLayerCacheSymbol];
};
var getB2BDataLayer = function getB2BDataLayer(publicToken, options) {
  var dataLayerCache = getB2BDataLayerCache();
  if (!dataLayerCache[publicToken]) {
    dataLayerCache[publicToken] = new B2BSubscriptionDataLayer(publicToken, options);
  }
  return dataLayerCache[publicToken];
};
var addSubscriber = function addSubscriber(collection, subscriber) {
  var uniqueId = Math.random().toString(36).slice(-10);
  collection[uniqueId] = subscriber;
  return function () {
    return delete collection[uniqueId];
  };
};
var notifySubscribers = function notifySubscribers(collection, value) {
  Object.values(collection).forEach(function (cb) {
    return cb(value);
  });
};
var deepEqualData = createDeepEqual({
  KEYS_TO_EXCLUDE: ['last_accessed_at']
});
var BaseSubscriptionService = /*#__PURE__*/function () {
  function BaseSubscriptionService(_publicToken, _datalayer, _ref1) {
    var _this5 = this;
    var usingCustomApiEndpoint = _ref1.usingCustomApiEndpoint;
    _classCallCheck(this, BaseSubscriptionService);
    this._publicToken = _publicToken;
    this._datalayer = _datalayer;
    this.fromCache = true;
    this._listen = function (e) {
      if (e.key !== internal.getPersistentStorageKey(_this5._publicToken, '')) {
        return;
      }
      var parsedValue = e.newValue === null || e.newValue === 'null' ? null : JSON.parse(e.newValue);
      _this5.updateState(parsedValue, true);
    };
    window.addEventListener('storage', this._listen);
    if (!usingCustomApiEndpoint) {
      var _this$_datalayer$read = this._datalayer.readSessionCookie(),
        session_token = _this$_datalayer$read.session_token;
      if (!session_token) {
        this.destroySession();
        return;
      }
    }
  }
  return _createClass(BaseSubscriptionService, [{
    key: "getTokens",
    value: function getTokens() {
      var _this$_datalayer$read2 = this._datalayer.readSessionCookie(),
        session_token = _this$_datalayer$read2.session_token,
        session_jwt = _this$_datalayer$read2.session_jwt;
      if (!(typeof session_token === 'string') || !(typeof session_jwt === 'string')) {
        return null;
      }
      return {
        session_token: session_token,
        session_jwt: session_jwt
      };
    }
  }, {
    key: "getIntermediateSessionToken",
    value: function getIntermediateSessionToken() {
      return this._datalayer.readIntermediateSessionTokenCookie() || null;
    }
  }, {
    key: "destroyState",
    value: function destroyState() {
      this.updateStateAndTokens({
        state: null,
        session_token: null,
        session_jwt: null,
        intermediate_session_token: null
      });
    }
  }, {
    key: "destroySession",
    value: function destroySession() {
      this._datalayer.removeSessionCookie();
      this.updateState(null);
    }
  }, {
    key: "_updateStateAndTokensInternal",
    value: function _updateStateAndTokensInternal(stateDiff, options) {
      var oldStateValue = this._datalayer.state;
      var newStateValue = stateDiff.state === null ? null : Object.assign(Object.assign({}, this._datalayer.state), stateDiff.state);
      this._datalayer.state = newStateValue;
      var wasCached = this.getFromCache();
      this.setCacheRefreshed();
      if (wasCached || !deepEqualData(oldStateValue, newStateValue)) {
        var notification;
        if (newStateValue == null || options.fromCache) {
          notification = newStateValue;
        } else {
          notification = Object.assign(Object.assign({}, newStateValue), {
            sessionDurationMinutes: options.sessionDurationMinutes
          });
        }
        notifySubscribers(this._datalayer.subscriptions, notification);
      }
    }
  }, {
    key: "updateStateAndTokens",
    value: function updateStateAndTokens(stateDiff) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {
        fromCache: false
      };
      if (stateDiff.state) {
        if (typeof stateDiff.session_token === 'string') {
          this._datalayer.writeSessionCookie(stateDiff);
        } else {
          this._datalayer.removeSessionCookie();
        }
        this._datalayer.removeISTCookie();
      } else if (stateDiff.intermediate_session_token) {
        if (typeof stateDiff.intermediate_session_token === 'string') {
          this._datalayer.writeIntermediateSessionTokenCookie(stateDiff.intermediate_session_token);
        } else {
          this._datalayer.removeISTCookie();
        }
        this._datalayer.removeSessionCookie();
      } else {
        this._datalayer.removeSessionCookie();
        this._datalayer.removeISTCookie();
      }
      this._updateStateAndTokensInternal(stateDiff, options);
      this._datalayer.syncToLocalStorage();
    }
  }, {
    key: "updateState",
    value: function updateState(state) {
      var fromExternalSource = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
      var oldStateValue = this._datalayer.state;
      var newStateValue = state === null ? null : Object.assign(Object.assign({}, this._datalayer.state), state);
      this._datalayer.state = newStateValue;
      var wasCached = this.getFromCache();
      this.setCacheRefreshed();
      var hasStateChanged = !deepEqualData(oldStateValue, newStateValue);
      if (wasCached || hasStateChanged) {
        notifySubscribers(this._datalayer.subscriptions, newStateValue);
        if (hasStateChanged && !fromExternalSource) {
          this._datalayer.syncToLocalStorage();
        }
      }
    }
  }, {
    key: "updateTokens",
    value: function updateTokens(tokens) {
      var session_token = tokens.session_token,
        session_jwt = tokens.session_jwt;
      var cookie = this._datalayer.readSessionCookie();
      var diff = Object.assign(Object.assign({}, cookie), {
        session_token: session_token,
        session_jwt: session_jwt
      });
      if (typeof session_token === 'string' || typeof session_jwt === 'string') {
        this._datalayer.writeSessionCookie(diff);
        this._datalayer.removeISTCookie();
      } else {
        this._datalayer.removeSessionCookie();
      }
    }
  }, {
    key: "subscribeToState",
    value: function subscribeToState(callback) {
      return addSubscriber(this._datalayer.subscriptions, callback);
    }
  }, {
    key: "getState",
    value: function getState() {
      return this._datalayer.state;
    }
  }, {
    key: "destroy",
    value: function destroy() {
      window.removeEventListener('storage', this._listen);
    }
  }, {
    key: "syncFromDeviceStorage",
    value: function syncFromDeviceStorage() {
      return null;
    }
  }, {
    key: "getFromCache",
    value: function getFromCache() {
      return this.fromCache;
    }
  }, {
    key: "setCacheRefreshed",
    value: function setCacheRefreshed() {
      this.fromCache = false;
    }
  }]);
}();
var B2BSubscriptionService = /*#__PURE__*/function (_BaseSubscriptionServ) {
  function B2BSubscriptionService() {
    var _this6;
    _classCallCheck(this, B2BSubscriptionService);
    _this6 = _callSuper(this, B2BSubscriptionService, arguments);
    _this6.updateMember = function (member) {
      return _this6.updateState({
        member: member
      });
    };
    _this6.getMember = function () {
      var _a, _b;
      return (_b = (_a = _this6.getState()) === null || _a === void 0 ? void 0 : _a.member) !== null && _b !== void 0 ? _b : null;
    };
    _this6.updateOrganization = function (organization) {
      return _this6.updateState({
        organization: organization
      });
    };
    _this6.getOrganization = function () {
      var _a, _b;
      return (_b = (_a = _this6.getState()) === null || _a === void 0 ? void 0 : _a.organization) !== null && _b !== void 0 ? _b : null;
    };
    _this6.getSession = function () {
      var _a, _b;
      return (_b = (_a = _this6.getState()) === null || _a === void 0 ? void 0 : _a.session) !== null && _b !== void 0 ? _b : null;
    };
    _this6.updateSession = function (resp, options) {
      if ('member_session' in resp && resp.member_session) {
        var tokens = resp.session_token && resp.session_jwt ? {
          session_token: resp.session_token,
          session_jwt: resp.session_jwt
        } : {
          session_token: true,
          session_jwt: true
        };
        _this6.updateStateAndTokens(Object.assign(Object.assign({
          state: {
            session: resp.member_session,
            member: resp.member,
            organization: resp.organization
          }
        }, tokens), {
          intermediate_session_token: null
        }), {
          fromCache: false,
          sessionDurationMinutes: options === null || options === void 0 ? void 0 : options.sessionDurationMinutes
        });
      } else {
        _this6.updateStateAndTokens({
          state: null,
          session_token: null,
          session_jwt: null,
          intermediate_session_token: resp.intermediate_session_token || true
        }, {
          fromCache: false,
          sessionDurationMinutes: options === null || options === void 0 ? void 0 : options.sessionDurationMinutes
        });
      }
    };
    _this6.withUpdateSession = function (authenticate) {
      return function (options) {
        return internal.__awaiter(_this6, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee8() {
          var resp;
          return _regenerator().w(function (_context8) {
            while (1) switch (_context8.n) {
              case 0:
                _context8.n = 1;
                return authenticate(options);
              case 1:
                resp = _context8.v;
                this.updateSession(resp, {
                  sessionDurationMinutes: options === null || options === void 0 ? void 0 : options.session_duration_minutes
                });
                return _context8.a(2, resp);
            }
          }, _callee8, this);
        }));
      };
    };
    return _this6;
  }
  _inherits(B2BSubscriptionService, _BaseSubscriptionServ);
  return _createClass(B2BSubscriptionService);
}(BaseSubscriptionService);
var HeadlessB2BOAuthClient = /*#__PURE__*/function (_internal$HeadlessB2B) {
  function HeadlessB2BOAuthClient(_networkClient, _subscriptionService, _pkceManager, _dynamicConfig, _config) {
    var _this7;
    var _dfpProtectedAuth = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : internal.DisabledDFPProtectedAuthProvider();
    var _oneTap = arguments.length > 6 ? arguments[6] : undefined;
    _classCallCheck(this, HeadlessB2BOAuthClient);
    _this7 = _callSuper(this, HeadlessB2BOAuthClient, [_networkClient, _subscriptionService, _pkceManager, _dynamicConfig, _config, _dfpProtectedAuth]);
    _this7._oneTap = _oneTap;
    _this7.startOneTapDiscovery = function (options) {
      return internal.__awaiter(_this7, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee9() {
        var clientResult, client, onSuccessCallback;
        return _regenerator().w(function (_context9) {
          while (1) switch (_context9.n) {
            case 0:
              _context9.n = 1;
              return this._oneTap.createOneTapClient();
            case 1:
              clientResult = _context9.v;
              if (clientResult.success) {
                _context9.n = 2;
                break;
              }
              throw new Error("One Tap could not load: ".concat(clientResult.reason));
            case 2:
              client = clientResult.client;
              onSuccessCallback = this._oneTap.createOnDiscoverySuccessHandler({
                discoveryRedirectUrl: options.discovery_redirect_url,
                onSuccess: this._oneTap.redirectOnSuccess
              });
              return _context9.a(2, client.render({
                style: {
                  position: internal.OneTapPositions.floating
                },
                callback: onSuccessCallback,
                cancelOnTapOutside: options.cancel_on_tap_outside
              }));
          }
        }, _callee9, this);
      }));
    };
    _this7.startOneTap = function (options) {
      return internal.__awaiter(_this7, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee0() {
        var clientResult, client, onSuccessCallback;
        return _regenerator().w(function (_context0) {
          while (1) switch (_context0.n) {
            case 0:
              if (options.organization_id) {
                _context0.n = 1;
                break;
              }
              throw new internal.StytchSDKUsageError('stytch.oauth.googleOneTap.start', 'organization_id is a required argument');
            case 1:
              _context0.n = 2;
              return this._oneTap.createOneTapClient();
            case 2:
              clientResult = _context0.v;
              if (clientResult.success) {
                _context0.n = 3;
                break;
              }
              throw new Error("One Tap could not load: ".concat(clientResult.reason));
            case 3:
              client = clientResult.client;
              onSuccessCallback = this._oneTap.createOnSuccessHandler({
                organizationId: options.organization_id,
                signupRedirectUrl: options.signup_redirect_url,
                loginRedirectUrl: options.login_redirect_url,
                onSuccess: this._oneTap.redirectOnSuccess
              });
              return _context0.a(2, client.render({
                style: {
                  position: internal.OneTapPositions.floating
                },
                callback: onSuccessCallback,
                cancelOnTapOutside: options.cancel_on_tap_outside
              }));
          }
        }, _callee0, this);
      }));
    };
    _this7.googleOneTap = {
      discovery: {
        start: _this7.startOneTapDiscovery
      },
      start: _this7.startOneTap
    };
    return _this7;
  }
  _inherits(HeadlessB2BOAuthClient, _internal$HeadlessB2B);
  return _createClass(HeadlessB2BOAuthClient);
}(internal.HeadlessB2BOAuthClient);
var navigatorSupportsFedCM = typeof window !== 'undefined' && 'IdentityCredential' in window;
var getConfiguredEmbeddedMode = function getConfiguredEmbeddedMode(position) {
  if (position === internal.OneTapPositions.forceLegacyEmbedded) {
    return 'force';
  }
  if (position === internal.OneTapPositions.embeddedOnly || position === internal.OneTapPositions.embedded || position === internal.OneTapPositions.floatingOrEmbedded) {
    return true;
  }
  return false;
};
var getShouldRenderEmbeddedOneTap = function getShouldRenderEmbeddedOneTap(position) {
  var embeddedMode = getConfiguredEmbeddedMode(position);
  return embeddedMode === 'force' || embeddedMode && !navigatorSupportsFedCM;
};
var GoogleOneTapClient = /*#__PURE__*/function () {
  function GoogleOneTapClient(_googleClient, _clientId) {
    _classCallCheck(this, GoogleOneTapClient);
    this._googleClient = _googleClient;
    this._clientId = _clientId;
  }
  return _createClass(GoogleOneTapClient, [{
    key: "cancel",
    value: function cancel() {
      this._googleClient.cancel();
    }
  }, {
    key: "render",
    value: function render(_ref10) {
      var callback = _ref10.callback,
        onOneTapCancelled = _ref10.onOneTapCancelled,
        style = _ref10.style,
        _ref10$cancelOnTapOut = _ref10.cancelOnTapOutside,
        cancelOnTapOutside = _ref10$cancelOnTapOut === void 0 ? true : _ref10$cancelOnTapOut;
      return internal.__awaiter(this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee1() {
        var _this8 = this;
        var embeddedMode, shouldRenderEmbeddedOneTap, config;
        return _regenerator().w(function (_context1) {
          while (1) switch (_context1.n) {
            case 0:
              embeddedMode = getConfiguredEmbeddedMode(style.position);
              shouldRenderEmbeddedOneTap = getShouldRenderEmbeddedOneTap(style.position);
              config = {
                client_id: this._clientId,
                callback: callback,
                auto_select: false,
                context: 'use',
                itp_support: true,
                use_fedcm_for_prompt: embeddedMode !== 'force',
                cancel_on_tap_outside: cancelOnTapOutside
              };
              if (shouldRenderEmbeddedOneTap) {
                config.prompt_parent_id = 'google-parent-prompt';
                config.cancel_on_tap_outside = false;
              }
              this._googleClient.initialize(config);
              return _context1.a(2, new Promise(function (resolve) {
                _this8._googleClient.prompt(function (notification) {
                  var _a, _b, _c, _d, _e;
                  if (notification.isSkippedMoment()) {
                    var reason = (_b = (_a = notification.getSkippedReason) === null || _a === void 0 ? void 0 : _a.call(notification)) !== null && _b !== void 0 ? _b : 'unknown_reason';
                    if (reason === 'user_cancel') {
                      onOneTapCancelled === null || onOneTapCancelled === void 0 ? void 0 : onOneTapCancelled();
                    }
                    return resolve({
                      success: false,
                      reason: reason
                    });
                  }
                  if (!notification.isNotDisplayed && shouldRenderEmbeddedOneTap || ((_c = notification.isNotDisplayed) === null || _c === void 0 ? void 0 : _c.call(notification))) {
                    return resolve({
                      success: false,
                      reason: (_e = (_d = notification.getNotDisplayedReason) === null || _d === void 0 ? void 0 : _d.call(notification)) !== null && _e !== void 0 ? _e : 'unknown_reason'
                    });
                  }
                  if (!notification.isDismissedMoment()) {
                    _this8.styleFrame(shouldRenderEmbeddedOneTap);
                    return resolve({
                      success: true
                    });
                  }
                });
              }));
          }
        }, _callee1, this);
      }));
    }
  }, {
    key: "styleFrame",
    value: function styleFrame(shouldRenderEmbeddedOneTap) {
      if (!shouldRenderEmbeddedOneTap) {
        return;
      }
      Array.from(document.getElementsByTagName('iframe')).forEach(function (frame) {
        if (frame.src.includes(internal.GOOGLE_ONE_TAP_HOST)) {
          frame.style.width = '100%';
        }
      });
    }
  }]);
}();
var loadGoogleOneTapClient = function loadGoogleOneTapClient() {
  return internal.loadESModule(internal.GOOGLE_ONE_TAP_SCRIPT_URL, function () {
    return window.google.accounts.id;
  });
};
var DefaultDynamicConfig = Promise.resolve({
  pkceRequiredForOAuth: false
});
var B2BOneTapProvider = /*#__PURE__*/function () {
  function B2BOneTapProvider(_networkClient, _pkceManager) {
    var _this9 = this;
    var _config = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : DefaultDynamicConfig;
    _classCallCheck(this, B2BOneTapProvider);
    this._networkClient = _networkClient;
    this._pkceManager = _pkceManager;
    this._config = _config;
    this.createOnDiscoverySuccessHandler = function (_ref11) {
      var discoveryRedirectUrl = _ref11.discoveryRedirectUrl,
        onSuccess = _ref11.onSuccess,
        onError = _ref11.onError;
      return function (response) {
        return internal.__awaiter(_this9, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee10() {
          var credential, submitPromise;
          return _regenerator().w(function (_context10) {
            while (1) switch (_context10.n) {
              case 0:
                credential = response.credential;
                submitPromise = this.submitGoogleOneTapTokenDiscovery({
                  credential: credential,
                  discoveryRedirectUrl: discoveryRedirectUrl
                }).then(function (result) {
                  return onSuccess(result.redirect_url);
                });
                if (onError) {
                  submitPromise = submitPromise["catch"](function (error) {
                    return onError(error);
                  });
                }
                _context10.n = 1;
                return submitPromise;
              case 1:
                return _context10.a(2);
            }
          }, _callee10, this);
        }));
      };
    };
    this.createOnSuccessHandler = function (_ref12) {
      var organizationId = _ref12.organizationId,
        signupRedirectUrl = _ref12.signupRedirectUrl,
        loginRedirectUrl = _ref12.loginRedirectUrl,
        onSuccess = _ref12.onSuccess,
        onError = _ref12.onError;
      return function (response) {
        return internal.__awaiter(_this9, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee11() {
          var credential, submitPromise;
          return _regenerator().w(function (_context11) {
            while (1) switch (_context11.n) {
              case 0:
                credential = response.credential;
                submitPromise = this.submitGoogleOneTapToken({
                  credential: credential,
                  organizationId: organizationId,
                  signupRedirectUrl: signupRedirectUrl,
                  loginRedirectUrl: loginRedirectUrl
                }).then(function (result) {
                  return onSuccess(result.redirect_url);
                });
                if (onError) {
                  submitPromise = submitPromise["catch"](function (error) {
                    return onError(error);
                  });
                }
                _context11.n = 1;
                return submitPromise;
              case 1:
                return _context11.a(2);
            }
          }, _callee11, this);
        }));
      };
    };
    this.redirectOnSuccess = function (redirect_url) {
      window.location.href = redirect_url;
    };
  }
  return _createClass(B2BOneTapProvider, [{
    key: "fetchGoogleStart",
    value: function fetchGoogleStart() {
      return internal.__awaiter(this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee12() {
        var oneTapStartResp;
        return _regenerator().w(function (_context12) {
          while (1) switch (_context12.n) {
            case 0:
              if (!this.googleClientID) {
                _context12.n = 1;
                break;
              }
              return _context12.a(2, this.googleClientID);
            case 1:
              _context12.n = 2;
              return this._networkClient.fetchSDK({
                url: '/b2b/oauth/google/onetap/start',
                method: 'GET'
              });
            case 2:
              oneTapStartResp = _context12.v;
              this.googleClientID = oneTapStartResp.google_client_id;
              return _context12.a(2, this.googleClientID);
          }
        }, _callee12, this);
      }));
    }
  }, {
    key: "submitGoogleOneTapTokenDiscovery",
    value: function submitGoogleOneTapTokenDiscovery(_ref13) {
      var credential = _ref13.credential,
        discoveryRedirectUrl = _ref13.discoveryRedirectUrl;
      return internal.__awaiter(this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee13() {
        var codeChallenge, _yield$this$_config, pkceRequiredForOAuth, keyPair;
        return _regenerator().w(function (_context13) {
          while (1) switch (_context13.n) {
            case 0:
              codeChallenge = undefined;
              _context13.n = 1;
              return this._config;
            case 1:
              _yield$this$_config = _context13.v;
              pkceRequiredForOAuth = _yield$this$_config.pkceRequiredForOAuth;
              if (!pkceRequiredForOAuth) {
                _context13.n = 3;
                break;
              }
              _context13.n = 2;
              return this._pkceManager.startPKCETransaction();
            case 2:
              keyPair = _context13.v;
              codeChallenge = keyPair.code_challenge;
              _context13.n = 4;
              break;
            case 3:
              this._pkceManager.clearPKPair();
            case 4:
              _context13.n = 5;
              return this._networkClient.fetchSDK({
                url: '/b2b/oauth/google/onetap/discovery/submit',
                method: 'POST',
                body: {
                  id_token: credential,
                  discovery_redirect_url: discoveryRedirectUrl,
                  pkce_code_challenge: codeChallenge
                }
              });
            case 5:
              return _context13.a(2, _context13.v);
          }
        }, _callee13, this);
      }));
    }
  }, {
    key: "submitGoogleOneTapToken",
    value: function submitGoogleOneTapToken(_ref14) {
      var credential = _ref14.credential,
        organizationId = _ref14.organizationId,
        signupRedirectUrl = _ref14.signupRedirectUrl,
        loginRedirectUrl = _ref14.loginRedirectUrl;
      return internal.__awaiter(this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee14() {
        var codeChallenge, _yield$this$_config2, pkceRequiredForOAuth, keyPair;
        return _regenerator().w(function (_context14) {
          while (1) switch (_context14.n) {
            case 0:
              codeChallenge = undefined;
              _context14.n = 1;
              return this._config;
            case 1:
              _yield$this$_config2 = _context14.v;
              pkceRequiredForOAuth = _yield$this$_config2.pkceRequiredForOAuth;
              if (!pkceRequiredForOAuth) {
                _context14.n = 3;
                break;
              }
              _context14.n = 2;
              return this._pkceManager.startPKCETransaction();
            case 2:
              keyPair = _context14.v;
              codeChallenge = keyPair.code_challenge;
              _context14.n = 4;
              break;
            case 3:
              this._pkceManager.clearPKPair();
            case 4:
              _context14.n = 5;
              return this._networkClient.fetchSDK({
                url: '/b2b/oauth/google/onetap/submit',
                method: 'POST',
                body: {
                  id_token: credential,
                  organization_id: organizationId,
                  signup_redirect_url: signupRedirectUrl,
                  login_redirect_url: loginRedirectUrl,
                  pkce_code_challenge: codeChallenge
                }
              });
            case 5:
              return _context14.a(2, _context14.v);
          }
        }, _callee14, this);
      }));
    }
  }, {
    key: "createOneTapClient",
    value: function createOneTapClient() {
      return internal.__awaiter(this, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee15() {
        var googleClientId, err, client, _t3, _t4, _t5, _t6;
        return _regenerator().w(function (_context15) {
          while (1) switch (_context15.p = _context15.n) {
            case 0:
              _context15.p = 0;
              _context15.n = 1;
              return this.fetchGoogleStart();
            case 1:
              googleClientId = _context15.v;
              _context15.n = 3;
              break;
            case 2:
              _context15.p = 2;
              _t3 = _context15.v;
              err = internal.StytchAPIError.from(_t3);
              return _context15.a(2, {
                success: false,
                reason: err.error_type
              });
            case 3:
              if (!(googleClientId === '')) {
                _context15.n = 4;
                break;
              }
              return _context15.a(2, {
                success: false,
                reason: 'oauth_config_not_found'
              });
            case 4:
              _t4 = GoogleOneTapClient;
              _context15.n = 5;
              return loadGoogleOneTapClient();
            case 5:
              _t5 = _context15.v;
              _t6 = googleClientId;
              client = new _t4(_t5, _t6);
              return _context15.a(2, {
                success: true,
                client: client
              });
          }
        }, _callee15, this, [[0, 2]]);
      }));
    }
  }]);
}();
var buildFinalConfig = function buildFinalConfig(clientName) {
  var opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var _a, _b;
  var cookieOptions = opts.cookieOptions,
    keepSessionAlive = opts.keepSessionAlive;
  var validator = internal.validate(clientName);
  validator.isOptionalBoolean('keepSessionAlive', keepSessionAlive);
  if (cookieOptions) {
    validator.isOptionalString('cookieOptions.opaqueTokenCookieName', cookieOptions.opaqueTokenCookieName).isOptionalString('cookieOptions.jwtCookieName', cookieOptions.jwtCookieName).isOptionalString('cookieOptions.istCookieName', cookieOptions.istCookieName).isOptionalString('cookieOptions.path', cookieOptions.path).isOptionalString('cookieOptions.domain', cookieOptions.domain);
  }
  return {
    cookieOptions: cookieOptions,
    keepSessionAlive: keepSessionAlive,
    endpoints: {
      testAPIURL: internal.getTestApiURL(opts),
      liveAPIURL: internal.getLiveApiURL(opts),
      dfpBackendURL: internal.getDFPBackendURL(opts),
      dfpCdnURL: internal.getDFPCdnURL(opts),
      clientsideServicesIframeURL: (_b = (_a = opts === null || opts === void 0 ? void 0 : opts.endpoints) === null || _a === void 0 ? void 0 : _a.clientsideServicesIframeURL) !== null && _b !== void 0 ? _b : internal.CLIENTSIDE_SERVICES_IFRAME_URL
    }
  };
};
var hasCustomApiEndpoint = function hasCustomApiEndpoint(token) {
  var rawOptions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var customBaseUrl = rawOptions.customBaseUrl,
    endpointOptions = rawOptions.endpointOptions;
  return internal.isTestPublicToken(token) ? !!(customBaseUrl || (endpointOptions === null || endpointOptions === void 0 ? void 0 : endpointOptions.testApiDomain)) : !!(customBaseUrl || (endpointOptions === null || endpointOptions === void 0 ? void 0 : endpointOptions.apiDomain));
};
function isHandledTokenType(handledTypes, tokenType) {
  return handledTypes.includes(tokenType);
}
var clearStytchTokenParams = function clearStytchTokenParams() {
  var url = removeStytchTokenParams(window.location.toString());
  window.history.replaceState(null, document.title, url);
};
var createAuthUrlHandler = function createAuthUrlHandler(handlers) {
  var handledTokenTypes = Object.keys(handlers);
  var parseAuthenticateUrl = function parseAuthenticateUrl() {
    var href = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : window.location.href;
    var url = new URL(href);
    var tokenType = url.searchParams.get('stytch_token_type');
    var token = url.searchParams.get('token');
    if (!token || !tokenType) {
      return null;
    }
    if (isHandledTokenType(handledTokenTypes, tokenType)) {
      return {
        handled: true,
        token: token,
        tokenType: tokenType
      };
    }
    return {
      handled: false,
      token: token,
      tokenType: tokenType
    };
  };
  var authenticateByUrl = function authenticateByUrl(_a, href) {
    return internal.__awaiter(void 0, void 0, void 0, /*#__PURE__*/_regenerator().m(function _callee16() {
      var clearParams, options, shouldClearParams, parsed, token, tokenType, handler, data;
      return _regenerator().w(function (_context16) {
        while (1) switch (_context16.p = _context16.n) {
          case 0:
            clearParams = _a.clearParams, options = internal.__rest(_a, ["clearParams"]);
            if (href === void 0) {
              href = window.location.href;
            }
            shouldClearParams = clearParams !== null && clearParams !== void 0 ? clearParams : href === window.location.href;
            parsed = parseAuthenticateUrl(href);
            if (!(parsed == null)) {
              _context16.n = 1;
              break;
            }
            return _context16.a(2, null);
          case 1:
            if (parsed.handled) {
              _context16.n = 2;
              break;
            }
            return _context16.a(2, parsed);
          case 2:
            token = parsed.token, tokenType = parsed.tokenType;
            handler = handlers[tokenType];
            _context16.p = 3;
            _context16.n = 4;
            return handler(token, options);
          case 4:
            data = _context16.v;
            return _context16.a(2, {
              handled: true,
              tokenType: tokenType,
              data: data
            });
          case 5:
            _context16.p = 5;
            if (shouldClearParams) {
              clearStytchTokenParams();
            }
            return _context16.f(5);
          case 6:
            return _context16.a(2);
        }
      }, _callee16, null, [[3,, 5, 6]]);
    }));
  };
  return {
    authenticateByUrl: authenticateByUrl,
    parseAuthenticateUrl: parseAuthenticateUrl
  };
};
function removeStytchTokenParams(href) {
  var url = new URL(href);
  var params = url.searchParams;
  params["delete"]('token');
  params["delete"]('stytch_token_type');
  return url;
}
var StytchB2BHeadlessClient = /*#__PURE__*/_createClass(function StytchB2BHeadlessClient(rawPublicToken, options) {
  var _this0 = this;
  _classCallCheck(this, StytchB2BHeadlessClient);
  internal.checkB2BNotSSR('StytchB2BHeadlessClient');
  internal.checkPublicToken(rawPublicToken);
  var _PUBLIC_TOKEN = rawPublicToken !== null && rawPublicToken !== void 0 ? rawPublicToken : '';
  var config = buildFinalConfig(this.constructor.name, options);
  var usingCustomApiEndpoint = hasCustomApiEndpoint(_PUBLIC_TOKEN, options);
  this._dataLayer = getB2BDataLayer(_PUBLIC_TOKEN, config);
  this._subscriptionService = new B2BSubscriptionService(_PUBLIC_TOKEN, this._dataLayer, {
    usingCustomApiEndpoint: usingCustomApiEndpoint
  });
  this._stateChangeClient = new internal.StateChangeClient(this._subscriptionService, {});
  this.onStateChange = function () {
    var _this0$_stateChangeCl;
    return (_this0$_stateChangeCl = _this0._stateChangeClient).onStateChange.apply(_this0$_stateChangeCl, arguments);
  };
  var additionalTelemetryDataFn = function additionalTelemetryDataFn() {
    var _a, _b, _c, _d;
    return {
      stytch_member_id: (_b = (_a = _this0._dataLayer.state) === null || _a === void 0 ? void 0 : _a.member) === null || _b === void 0 ? void 0 : _b.member_id,
      stytch_member_session_id: (_d = (_c = _this0._dataLayer.state) === null || _c === void 0 ? void 0 : _c.session) === null || _d === void 0 ? void 0 : _d.member_session_id
    };
  };
  var passwordsPKCEManager = new PKCEManager(this._dataLayer, 'passwords');
  var networkClient = new NetworkClient(_PUBLIC_TOKEN, this._dataLayer, config.endpoints.liveAPIURL, config.endpoints.testAPIURL, additionalTelemetryDataFn);
  this._networkClient = networkClient;
  var bootstrap = new BootstrapDataManager(_PUBLIC_TOKEN, networkClient, this._dataLayer);
  var captcha = new CaptchaProvider(bootstrap.getAsync());
  var dfpProtectedAuth = new internal.DFPProtectedAuthProvider(_PUBLIC_TOKEN, config.endpoints.dfpBackendURL, config.endpoints.dfpCdnURL, bootstrap.getAsync(), captcha.executeRecaptcha);
  var clientsideServices = new ClientsideServicesProvider(config.endpoints.clientsideServicesIframeURL);
  var pkceManagerForOAuth = new PKCEManager(this._dataLayer, 'oauth');
  var oneTap = new B2BOneTapProvider(networkClient, pkceManagerForOAuth, bootstrap.getAsync());
  this.organization = new internal.HeadlessB2BOrganizationClient(networkClient, networkClient, this._subscriptionService);
  this.member = new internal.HeadlessB2BSelfClient(networkClient, networkClient, this._subscriptionService);
  this.self = new internal.HeadlessB2BSelfClient(networkClient, networkClient, this._subscriptionService);
  this.session = new internal.HeadlessB2BSessionClient(networkClient, this._subscriptionService);
  this.magicLinks = new internal.HeadlessB2BMagicLinksClient(networkClient, this._subscriptionService, new PKCEManager(this._dataLayer, 'magic_links'), passwordsPKCEManager, bootstrap.getAsync(), dfpProtectedAuth);
  this.oauth = new HeadlessB2BOAuthClient(networkClient, this._subscriptionService, pkceManagerForOAuth, bootstrap.getAsync(), {
    publicToken: _PUBLIC_TOKEN,
    testAPIURL: config.endpoints.testAPIURL,
    liveAPIURL: config.endpoints.liveAPIURL
  }, dfpProtectedAuth, oneTap);
  this.sso = new internal.HeadlessB2BSSOClient(networkClient, this._subscriptionService, new PKCEManager(this._dataLayer, 'sso'), bootstrap.getAsync(), {
    publicToken: _PUBLIC_TOKEN,
    testAPIURL: config.endpoints.testAPIURL,
    liveAPIURL: config.endpoints.liveAPIURL
  }, dfpProtectedAuth);
  this.discovery = new internal.HeadlessB2BDiscoveryClient(networkClient, this._subscriptionService);
  this.passwords = new internal.HeadlessB2BPasswordsClient(networkClient, this._subscriptionService, passwordsPKCEManager, bootstrap.getAsync(), dfpProtectedAuth);
  this.otps = new internal.HeadlessB2BOTPsClient(networkClient, this._subscriptionService, dfpProtectedAuth);
  this.totp = new internal.HeadlessB2BTOTPsClient(networkClient, this._subscriptionService, dfpProtectedAuth);
  this.recoveryCodes = new internal.HeadlessB2BRecoveryCodesClient(networkClient, this._subscriptionService, dfpProtectedAuth);
  this.rbac = new internal.HeadlessB2BRBACClient(bootstrap.getSync(), bootstrap.getAsync(), this._subscriptionService);
  this.scim = new internal.HeadlessB2BSCIMClient(this._networkClient, this._subscriptionService);
  this.impersonation = new internal.HeadlessB2BImpersonationClient(this._networkClient, this._subscriptionService, dfpProtectedAuth);
  this.idp = new internal.HeadlessB2BIDPClient(networkClient);
  this._sessionManager = new internal.SessionManager(this._subscriptionService, this.session, _PUBLIC_TOKEN, {
    keepSessionAlive: config.keepSessionAlive
  });
  var searchManager = new internal.SearchDataManager(networkClient, dfpProtectedAuth);
  if (usingCustomApiEndpoint || this._dataLayer.readSessionCookie().session_token) {
    this._sessionManager.performBackgroundRefresh();
  }
  var _createAuthUrlHandler = createAuthUrlHandler({
      discovery: function discovery(token) {
        return _this0.magicLinks.discovery.authenticate({
          discovery_magic_links_token: token
        });
      },
      discovery_oauth: function discovery_oauth(token) {
        return _this0.oauth.discovery.authenticate({
          discovery_oauth_token: token
        });
      },
      oauth: function oauth(token, options) {
        return _this0.oauth.authenticate(Object.assign({
          oauth_token: token
        }, options));
      },
      sso: function sso(token, options) {
        return _this0.sso.authenticate(Object.assign({
          sso_token: token
        }, options));
      },
      multi_tenant_magic_links: function multi_tenant_magic_links(token, options) {
        return _this0.magicLinks.authenticate(Object.assign({
          magic_links_token: token
        }, options));
      },
      multi_tenant_impersonation: function multi_tenant_impersonation(token) {
        return _this0.impersonation.authenticate({
          impersonation_token: token
        });
      }
    }),
    authenticateByUrl = _createAuthUrlHandler.authenticateByUrl,
    parseAuthenticateUrl = _createAuthUrlHandler.parseAuthenticateUrl;
  this.authenticateByUrl = authenticateByUrl;
  this.parseAuthenticateUrl = parseAuthenticateUrl;
  networkClient.logEvent({
    name: 'b2b_sdk_instance_instantiated',
    details: {
      event_callback_registered: false,
      error_callback_registered: false,
      success_callback_registered: false
    }
  });
  bootstrap.getAsync().then(function (bootstrapData) {
    if (bootstrapData.vertical === internal.VERTICAL_CONSUMER) {
      internal.logger.error('This application is using a Stytch client for B2B projects, but the public token is for a Stytch Consumer project. Use a Consumer Stytch client instead, or verify that the public token is correct.');
    }
  });
  var internals = {
    bootstrap: bootstrap,
    clientsideServices: clientsideServices,
    publicToken: _PUBLIC_TOKEN,
    searchManager: searchManager,
    dataLayer: this._dataLayer,
    networkClient: networkClient,
    oneTap: oneTap
  };
  internal.writeB2BInternals(this, internals);
});
exports.StytchB2BHeadlessClient = StytchB2BHeadlessClient;
exports.clearStytchTokenParams = clearStytchTokenParams;
