// @stytch/vanilla-js/dist/b2b/index.headless.js
'use strict';

var b2b_index_headless = require('../index.headless-Dj4KRdHg.js');
var internal = require('../internal-BWEZsr7G.js');
exports.StytchB2BHeadlessClient = b2b_index_headless.StytchB2BHeadlessClient;
Object.defineProperty(exports, "AuthFlowType", {
  enumerable: true,
  get: function get() {
    return internal.AuthFlowType;
  }
});
Object.defineProperty(exports, "B2BMFAProducts", {
  enumerable: true,
  get: function get() {
    return internal.B2BMFAProducts;
  }
});
Object.defineProperty(exports, "B2BOAuthProviders", {
  enumerable: true,
  get: function get() {
    return internal.B2BOAuthProviders;
  }
});
Object.defineProperty(exports, "B2BProducts", {
  enumerable: true,
  get: function get() {
    return internal.B2BProducts;
  }
});
exports.BiometricRegistrationIdIsNullOrBlank = internal.BiometricRegistrationIdIsNullOrBlank;
exports.BiometricsAlreadyEnrolledError = internal.BiometricsAlreadyEnrolledError;
exports.BiometricsFailedError = internal.BiometricsFailedError;
exports.BiometricsUnavailableError = internal.BiometricsUnavailableError;
exports.ChallengeSigningFailedError = internal.ChallengeSigningFailedError;
exports.DFPNotConfigured = internal.DFPNotConfigured;
exports.DeviceCredentialsNotAllowedError = internal.DeviceCredentialsNotAllowedError;
exports.FailedCodeChallengeError = internal.FailedCodeChallengeError;
exports.FailedToDecryptDataError = internal.FailedToDecryptDataError;
exports.IDPOAuthFlowMissingParamError = internal.IDPOAuthFlowMissingParamError;
exports.InternalError = internal.InternalError;
exports.InvalidAuthorizationCredentialError = internal.InvalidAuthorizationCredentialError;
exports.InvalidCredentialTypeError = internal.InvalidCredentialTypeError;
exports.InvalidPrivateKeyLength = internal.InvalidPrivateKeyLength;
exports.InvalidRedirectSchemeError = internal.InvalidRedirectSchemeError;
exports.InvalidStartUrlError = internal.InvalidStartUrlError;
exports.JSONDataNotConvertibleToStringError = internal.JSONDataNotConvertibleToStringError;
exports.KeyInvalidatedError = internal.KeyInvalidatedError;
exports.KeystoreUnavailableError = internal.KeystoreUnavailableError;
exports.MissingAttestationObjectError = internal.MissingAttestationObjectError;
exports.MissingAuthorizationCredentialIDTokenError = internal.MissingAuthorizationCredentialIDTokenError;
exports.MissingCipherIv = internal.MissingCipherIv;
exports.MissingGoogleClientIDError = internal.MissingGoogleClientIDError;
exports.MissingPKCEError = internal.MissingPKCEError;
exports.MissingPublicKeyError = internal.MissingPublicKeyError;
exports.MissingUrlError = internal.MissingUrlError;
exports.NoBiometricsEnrolledError = internal.NoBiometricsEnrolledError;
exports.NoBiometricsRegistrationError = internal.NoBiometricsRegistrationError;
exports.NoCredentialsPresentError = internal.NoCredentialsPresentError;
exports.NoCurrentSessionError = internal.NoCurrentSessionError;
Object.defineProperty(exports, "OAuthProviders", {
  enumerable: true,
  get: function get() {
    return internal.OAuthProviders;
  }
});
Object.defineProperty(exports, "OTPMethods", {
  enumerable: true,
  get: function get() {
    return internal.OTPMethods;
  }
});
Object.defineProperty(exports, "OneTapPositions", {
  enumerable: true,
  get: function get() {
    return internal.OneTapPositions;
  }
});
exports.PasskeysInvalidEncoding = internal.PasskeysInvalidEncoding;
exports.PasskeysMisconfigured = internal.PasskeysMisconfigured;
exports.PasskeysUnsupportedError = internal.PasskeysUnsupportedError;
Object.defineProperty(exports, "Products", {
  enumerable: true,
  get: function get() {
    return internal.Products;
  }
});
Object.defineProperty(exports, "RNUIProducts", {
  enumerable: true,
  get: function get() {
    return internal.RNUIProducts;
  }
});
exports.RandomNumberGenerationFailed = internal.RandomNumberGenerationFailed;
Object.defineProperty(exports, "RedirectURLType", {
  enumerable: true,
  get: function get() {
    return internal.RedirectURLType;
  }
});
exports.SDKAPIUnreachableError = internal.SDKAPIUnreachableError;
exports.SDKNotConfiguredError = internal.SDKNotConfiguredError;
exports.SignInWithAppleMisconfigured = internal.SignInWithAppleMisconfigured;
exports.StytchAPIError = internal.StytchAPIError;
exports.StytchAPISchemaError = internal.StytchAPISchemaError;
exports.StytchAPIUnreachableError = internal.StytchAPIUnreachableError;
exports.StytchError = internal.StytchError;
Object.defineProperty(exports, "StytchEventType", {
  enumerable: true,
  get: function get() {
    return internal.StytchEventType;
  }
});
exports.StytchSDKAPIError = internal.StytchSDKAPIError;
exports.StytchSDKError = internal.StytchSDKError;
exports.StytchSDKSchemaError = internal.StytchSDKSchemaError;
exports.StytchSDKUsageError = internal.StytchSDKUsageError;
exports.UNRECOVERABLE_ERROR_TYPES = internal.UNRECOVERABLE_ERROR_TYPES;
exports.UserCancellationError = internal.UserCancellationError;
exports.UserLockedOutError = internal.UserLockedOutError;
Object.defineProperty(exports, "Wallets", {
  enumerable: true,
  get: function get() {
    return internal.Wallets;
  }
});
exports.errorToStytchError = internal.errorToStytchError;
